### ISSUE TITLE
Guessing number with non-iid attempts

### ISSUE THREAD

**Issue by ShayanTalaei** [GnarlyMshtep/pass_at_k#46]
**One sentence**
From a set of numbers, the model has to guess a secret chosen from a distribution over the set.

**What is interesting to measure?**
Pass@k coverage of the secret number.

**What do you expect to see?**
The non-iid parallel sampling learns to explore different paths in different attempt ids.

**Comment by ShayanTalaei** [comment]
First version, the set of numbers is fixed {1,2,3,4}, and the secret number is chosen uniformly at random.

<img width="631" height="313" alt="Image" src="https://github.com/user-attachments/assets/e313dbd5-4680-4fce-8aaf-441561499f49" />

<img width="635" height="316" alt="Image" src="https://github.com/user-attachments/assets/0e39f959-6a7e-4ecc-a412-c7405faab82c" />


- The model struggles to learn the optimal strategy. 
- The grad norm converges to 0 very quickly.

**Hypothesis on why this happens?**

We have conflicting prompts in different train samples. In one example, attempt 1, might have guessed 1 and gotten rewarded, while in another, for the same guess, it gets penalized. 

Potential solution: Diversifying the prompts.

**Comment by ShayanTalaei** [comment]
**Data:** Selecting 4 numbers from {1, ..., 100} at random like {A, B, C, D} to be the set of options and then choosing one uniformly at random as the secret number for the task.

<img width="1274" height="313" alt="Image" src="https://github.com/user-attachments/assets/ffe06edd-7705-426d-8324-43d6a4aa2aad" />

<img width="628" height="312" alt="Image" src="https://github.com/user-attachments/assets/e780f942-b66b-4f87-9c1e-63aef620a25b" />

<img width="1268" height="318" alt="Image" src="https://github.com/user-attachments/assets/c74e673f-4ef5-421e-aa16-5bc269bb81e6" />

- Both the 3B and 7B models learn the task.
- The 7B model doesn't output the thinking and just learns to do the task. 
- The 3B model also learns the task after some point, and in that moment, its grad norm becomes 0. The entropy starts to increase (not sure why?).
- Non-iid strategy reaches 100%, while iid reaches ~72% which is the value that we can show is the optimal value for the iid in this case.

**Next**: Having more options than attempts.

**Comment by ShayanTalaei** [comment]
**Data:** Same as above but we first select 8 numbers from {1, ..., 100} as our set. The model still gets 4 attempts.

<img width="1200" height="314" alt="Image" src="https://github.com/user-attachments/assets/ec266cf9-59f7-4031-b189-05285590acf1" />

<img width="1811" height="629" alt="Image" src="https://github.com/user-attachments/assets/bad38634-e1bb-4ad2-9835-4ae9669c8869" />

<img width="1212" height="315" alt="Image" src="https://github.com/user-attachments/assets/5a7d587e-af96-491f-a476-a04ad46c362e" />

- Learning is very unstable and collapses after some point.

**Hypothesis on why this happens?**
- For the 7B model, it still doesn't think, and gets noisy signals.
- **Important**: the gap between the optimal non-iid and iid strategy becomes small as we increase N for a fixed $K_{\text{train}}$. This provides less signal for the model to learn from.

**How to solve this?**
- Some stuff for making the training more stable: reducing LR, increasing batch size, reducing the clipping thresholds, (?)
- Choosing the secret number with a non-uniform distribution from the set. This increases the gap between the iid and non-iid strategies.

**Comment by ShayanTalaei** [comment]
**Data:** Reducing the target set size back to 4, but choosing the secret number using an exponential distribution (option 1, with prob p, option 2 with prob p^2, ... all normalized to sum to 1). For this experiment, p is chosen to be 1/2.

<img width="1199" height="309" alt="Image" src="https://github.com/user-attachments/assets/e9ae05f1-93a5-48af-9fe0-54ea76d7934e" />

<img width="596" height="311" alt="Image" src="https://github.com/user-attachments/assets/e65bd6a4-7ed7-453e-8ee0-189bfd02bce9" />

<img width="1202" height="313" alt="Image" src="https://github.com/user-attachments/assets/9e64d3d6-442f-442a-8508-c8ccd8f461f2" />

- The results are very similar to the uniform case, and confirm that the model can learn the non-iid strategy even if the distribution is non-uniform.

**Next:** Non-uniform distribution with a set size of 8.

**Comment by ShayanTalaei** [comment]
**Data**: A set of 8 numbers is chosen from {1, ..., 100} assume {A1 < A2 < ...< A8}. A secret number X is chosen from this set with probability p^i being Ai. 

The model has 4 attempts to guess the secret number. The optimal non-iid strategy would be to guess the first 4 numbers of the set in the first 4 attempts which should bring ~1/2+1/4+1/8+1/16=1-1/16=0.9375

<img width="1204" height="312" alt="Image" src="https://github.com/user-attachments/assets/64277f7d-1e39-427d-bf76-aba2a60a0814" />

<img width="599" height="310" alt="Image" src="https://github.com/user-attachments/assets/ec06b60d-bd83-47b0-ba48-784e51a2766f" />

<img width="1202" height="314" alt="Image" src="https://github.com/user-attachments/assets/586a046d-23a3-4e68-a169-28f0d2e4cd7d" />

- The 7B model collapses and doesn't learn how to do the task.
- It **doesn't think*. We need to fix this.

**Next**: We should have the model to think about the fact that it's using its attempt i and somehow use that in its action.

**Comment by ShayanTalaei** [comment]
**Attempt template**:
`\n Come up with a list of high-potential strategies, enumerate strategies by their possibility, at attempt {attempt_id}, pick the {attempt_id}-th strategy. This is your attempt {attempt_id} out of {max_attempts} attempts."`

Maybe this is too explicit? @andrewilyas @GnarlyMshtep What do you think?

**Data:** Same as above, once with p=0.5, and once with p=0.9. Note that the second one is an almost uniform distribution that has a smaller gap between the iid and non-iid strategies.

<img width="1205" height="311" alt="Image" src="https://github.com/user-attachments/assets/c88ebd47-96cd-4c1d-b077-c2970b319ba6" />

<img width="1209" height="319" alt="Image" src="https://github.com/user-attachments/assets/69647ce8-9fc7-48fb-a17b-ad6626d1f225" />

<img width="1203" height="312" alt="Image" src="https://github.com/user-attachments/assets/dc741f2a-3ba1-4fd0-a519-3e680263d834" />



An example prompt and response:
**input**: 
`system\nYou are a helpful assistant. You will be given a task to guess a hidden number from a provided set of candidate numbers. You will be asked to answer the question a couple of times, and you will win if you guess the number correctly in at least one of your attempts. Pay attention to your attempt number, and the total number of attempts you have. Therefore, you should diversify your attempts, and devise a specific strategy considering your current attempt number and the total number of attempts you have. Think step by step between <think> and </think> and then provide your answer inside <answer> YOUR GUESS HERE </answer> tags as a single integer. Don't output any other text after the </answer>.\nuser\nI have chosen a secret number from the following set of numbers (all are distinct): 11, 12, 39, 52, 77, 83, 86, 98. Can you guess which one it is? Please think step by step between <think> and </think> and then provide your final answer inside <answer> YOUR GUESS HERE </answer> tags as a single integer.\n Come up with a list of high-potential strategies, enumerate strategies by their possibility, at attempt 3, pick the 3-th strategy. This is your attempt 3 out of 4 attempts.\nassistant\n"`

**"output"**: 
`"<think>\n1. The number could be any of the given numbers, so we need to consider a systematic approach.\n2. Since the numbers are distinct and there's no additional information, we can use a simple strategy like guessing in ascending order.\n3. Let's list the numbers in ascending order for consistency: 11, 12, 39, 52, 77, 83, 86, 98.\n4. Since we're on attempt 3, we'll pick the 3rd number in this list.\n</think>\n<answer> 39 </answer>"`

---
**Heatmaps of the model guesses**


<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/241eae98-74b8-4bbe-a28a-76e3a70c0869" />

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/56943ce4-6cd4-4477-b39a-080fc4575d3f" />

---
**Take-aways**
- Things are working! 🎊  
- In each case, the non-iid strategy both learn the optimal strategy verbally and achieves the optimal value that it should.
- In the almost-uniform case, the model learns to only output the answer without thinking about the strategy. Not sure if this is a good/bad thing? But, definitely interesting!

**Next**
- What if we reverse/shuffle the numbers when presented to the model? The strategy that it should learn is not as obvious as attempt i -> i'th number anymore.

**Comment by ShayanTalaei** [comment]
Before adding the explicit prompt, the model wasn't learning to do what we wanted. This can be simply due to the fact that attending to the attempt ID is not something that the model would do. More generally, if we want the RL to amplify a behavior, it should exist with some non-zero probability in the base model, either by fine-tuning, or prompting, or ?. 

Before making the prompt very explicit, we saw that the model is not converging (at least within the defined steps) to a high performance. 

**Next**: I'll test some cases where the described strategy doesn't exactly match what the model should do.

**Comment by ShayanTalaei** [comment]
**DATA**: 8 numbers from 1-100, and the distribution follows a power law (r=0.5) with the largest number having the highest probability. Very similar to the experiment above, with the difference that the most likely answer is the last one. 

<img width="486" height="311" alt="Image" src="https://github.com/user-attachments/assets/2a21b56a-b5b1-4627-a4ad-3bcf548259cc" />

<img width="732" height="630" alt="Image" src="https://github.com/user-attachments/assets/488e8e06-7e9d-4560-9d74-0880c08bf1da" />

<img width="730" height="632" alt="Image" src="https://github.com/user-attachments/assets/410ca50e-bd7c-47ce-b9ab-5af14e5d1425" />

<img width="733" height="315" alt="Image" src="https://github.com/user-attachments/assets/5925008b-c337-430b-8e01-ee9cfbc46a9c" />

**Take-aways**
- The run starts to converge initially. Around step 30, it reaches an almost zero gradient (at that point, it has achieved the optimal score of 1/2+1/4+1/8+1/16 on the batch). After that, it collapses.
- Not sure why and how to make the run more stable?

**Comment by ShayanTalaei** [comment]
**Data**: Guessing a number from a set of 4 implicit numbers.

Example:

**input**: 
`"system\nYou are a helpful assistant. You will be given a task to guess a hidden number from a set that is **described**, not listed explicitly. You will be asked to answer the question a couple of times, and you will win if you guess the number correctly in at least one of your attempts. Pay attention to your attempt number, and the total number of attempts you have. Therefore, you should diversify your attempts, and devise a specific strategy considering your current attempt number and the total number of attempts you have. Think step by step between <think> and </think> and then provide your answer inside <answer> YOUR GUESS HERE </answer> tags as a single integer. Don't output any other text after the </answer>.\nuser\nI have chosen a secret number from the following described set: Map letters using A=0, B=1, …, Z=25. Take Base = 646. Consider Base + {O, X, I, T} (setwise addition). Can you guess which one it is? Please think step by step between <think> and </think> and then provide your final answer inside <answer> YOUR GUESS HERE </answer> tags as a single integer.\n Come up with a list of high-potential strategies, enumerate strategies by their possibility, at attempt 1, pick the 1-th strategy. This is your attempt 1 out of 4 attempts.\nassistant\n"`

**output**: 
`"<think>\n1. The set is Base + {O, X, I, T}, where Base = 646 and A=0, B=1, ..., Z=25.\n2. O=14, X=23, I=8, T=19.\n3. Adding these to Base:\n   - O: Base + O = 646 + 14 = 660.\n   - X: Base + X = 646 + 23 = 669.\n   - I: Base + I = 646 + 8 = 654.\n   - T: Base + T = 646 + 19 = 665.\n4. The set of secret numbers is {660, 669, 654, 665}.\n</think>\n<answer> 660 </answer>"`

---

<img width="731" height="626" alt="Image" src="https://github.com/user-attachments/assets/0d0029ad-b285-4ff8-b9c2-47087ee9f58d" />

<img width="732" height="626" alt="Image" src="https://github.com/user-attachments/assets/4f41567f-8e41-4118-8917-cb528d17b3e5" />

<img width="733" height="633" alt="Image" src="https://github.com/user-attachments/assets/7e18cfbe-5d33-463c-b8f8-b1b8748719d7" />

<img width="731" height="313" alt="Image" src="https://github.com/user-attachments/assets/864fb8cd-e317-43e9-a738-6869bc2d5432" />

**Take-aways**
- The model seems to learn the task initially, but collapses after some point. @GnarlyMshtep @andrewilyas Any thoughts on what might be the cause and how to fix it?

**Comment by GnarlyMshtep** [comment]
Thanks a ton for this! Very informative. Questions: 
* what's `no_grad_norm` in the experiments? 
* are all of these with the variance across groups stabilization? (I'm assuming yes based on the order of ur messages)
* So it is working, reaches optimal reward, and then collapses? Really? That's very strange. 
* In the future, could u maybe post sc of plots u think are especially rel (maybe w a hint on what we should be looking for) and link to wandb run? This way we can check for additional info, e.g. cfg
* What do the rollouts look like in the step where it goes crazy? It basically looks to me like training is fine, there is a single crazy step, and the model collapses right after, e.g. see response length spike to $\infty$. 
* Are you using KL penalty? Policy clipping? The facy yjay kl spikes (in the orange plots and clipfrac doesn't might mean its too small?

**Comment by GnarlyMshtep** [comment]
@ShayanTalaei another thing u could try to understand / remove the spike in negative advantage is to set 
```python 
norm_adv_by_std_in_grpo = self.config.algorithm.get(
                            "norm_adv_by_std_in_grpo", True
                        )
```
to `False`. The reason is that I suspect that we have 0 advantage / close to 0 advantage + small score due to {KL, slight off-policy due to verl, entropy bonus, clipping on off-policy mini batch size, some other added config score idk} and it has small effect in magnitude and gets blown out of proportion by this. 

I don't think its a replacement to putting. a breakpoint in `fit()` after advantages are assigned to inspect all the variables in the step that went crazy, but its something to try.

**Comment by GnarlyMshtep** [comment]
We saw that things are relatively broken: 
* the steps that were destablizing indeed had all the same scores. 
* one thing I am suspicious of is that the reward is wrong for some of the set? (actually, it looks that its just not sufficently "stochastic" and that if the answer is not one of the most likely numbers according to the distribution, the LM won't guess it). 

We can implement one of the following mitigation measures
* don't backprop on all the same reward mini-batch by doing DAPO 

Another interesting observation is that the model only uses its 4 attempts to guess the top 3 numbers, because guessing the 4-th largest and it being correct is rare.

**Comment by GnarlyMshtep** [comment]
Meeting with @ShayanTalaei  right now. 
* he showed me pres for his advisor 
* he showed me experimental results (@ShayanTalaei  -- can u drop some wandb links here after?)
* he had dramatic takeaway that we need to rethink our reward function. The reason for this is that we were getting "lazy rollouts" where if, e.g. all attempt 1 are correct, then even if attempt 2 has some corrects and some wrongs, leaning won't happen. 

My vibe: 
* we kinda lack representative runs to learn from -- the base countdown run from yesterday just collapsed. 
* we should train 1.5B models, not 7B, since we don't need 7B intelligence and we can easily get 8x the runs / GPU power + faster 
* I think the best way to deal with the lazy agent problem is to alternate pass@1 and pass@k training. I think this is the best way because that is what the bytedance paper did and it worked well for them. 
* I think its also reasonable to try some convex combination of rewards, 1000 things which make sense to me, for example giving the `our_attempt_advantage * \lambda + is_correct_advantage * (1 - \lambda)` and we can also try this. 

Action items: 
* change dataset: only 1 epoch, larger dataset. 
* first train pass@1 for some steps (maybe w attempt but don't give attempt reward), get checkpoint, and train both ours and pass@1. 
* gradual increase, stay stable, gradual decrease, stay stable. (get current step number from `wandb`) -- can start with fixed $\lambda$, or switch.
